tests/test_discrete_quadratic_model.py

```diff
@@ -107,6 +139,96 @@ def test_two_variable_labelled(self):
         self.assertEqual(dqm.energy({u: 0, v: 1}), 2.5)
 
 
+class TestFile(unittest.TestCase):
+    def test_empty_functional(self):
+        dqm = dimod.DQM()
+
+        with tempfile.TemporaryFile() as tf:
+            with dqm.to_file() as df:
+                shutil.copyfileobj(df, tf)
+            tf.seek(0)
+            new = dimod.DQM.from_file(tf)
+
+        self.assertEqual(new.num_variables(), 0)
+
+    def test_two_var_functional(self):
+        dqm = dimod.DQM()
+        dqm.add_variable(5)
+        dqm.add_variable(7)
+
+        dqm.set_linear_case(0, 3, 1.5)
+        dqm.set_quadratic(0, 1, {(0, 1): 1.5, (3, 4): 1})
+
+        with tempfile.TemporaryFile() as tf:
+            with dqm.to_file() as df:
+                shutil.copyfileobj(df, tf)
+            tf.seek(0)
+            new = dimod.DQM.from_file(tf)
+
+        self.assertEqual(new.num_variables(), dqm.num_variables())
+        self.assertEqual(new.num_cases(), dqm.num_cases())
+        self.assertEqual(new.variables, dqm.variables)
+        for v in dqm.variables:
+            self.assertEqual(new.num_cases(v), dqm.num_cases(v))
+            np.testing.assert_array_equal(new.get_linear(v),
+                                          dqm.get_linear(v))
+        self.assertEqual(new.adj, dqm.adj)
```

Comment:
`DQM.__eq__`?

Reply:
The problem comes with handling differently ordered DQMs. Right now `bqm == other_bqm` returns `True` even when they have a different variable order. I would probably want to be consistent with BQM, but I wanted to think about it a bit more first. In either case this (mostly) tests the stronger condition that they are equal and have identical underlying values.